text_generation/causal_lm/cpp/generate_pipeline/generate_sample.cpp

```diff
@@ -0,0 +1,113 @@
+
+
+// The following reasons require TextStreamer to keep a cache of previous tokens:
+// detokenizer removes starting ' '. For example detokenize(tokenize(" a")) == "a",
+// but detokenize(tokenize("prefix a")) == "prefix a"
+// 1 printable token may consist of 2 token ids: detokenize(incomplete_token_idx) == "�"
+struct TextStreamer {
+    LLMPipeline pipe;
+    std::vector<int64_t> token_cache;
+    size_t print_len = 0;
+
+    void put(int64_t token) {
+        token_cache.push_back(token);
+        std::string text = pipe.detokenize(token_cache);
+        if (!text.empty() && '\n' == text.back()) {
+            // Flush the cache after the new line symbol
+            std::cout << std::string_view{text.data() + print_len, text.size() - print_len};
+            token_cache.clear();
+            print_len = 0;
+	    return;
+        }
+        if (text.size() >= 3 && text.compare(text.size() - 3, 3, "�") == 0) {
+            // Don't print incomplete text
+            return;
+        }
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << std::flush;
+        print_len = text.size();
+    }
+
+    void end() {
+        std::string text = pipe.detokenize(token_cache);
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << '\n';
+        token_cache.clear();
+        print_len = 0;
+    }
+};
+
+int main(int argc, char* argv[]) try {
+    if (2 >= argc && argc <= 4)
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> \"<PROMPT>\" <DEVICE>");
+    
+    std::string prompt = "table is made of";
+    std::string device = "CPU"; // can be replaced with GPU
+
+    std::string model_path = argv[1];
+    if (argc > 2)
+        prompt = argv[2];
+    if (argc > 3)
+        device = argv[3];
+
+    // Example 1: TextStreaming example with greedy search
+    LLMPipeline pipe(model_path, device);
+    // Will try to load config from generation_config.json.
+    // but if not found default velues for gready search will be used
+    GenerationConfig config = pipe.generation_config();
+
+    auto text_streamer = TextStreamer{pipe};
+    auto text_streamer_callback = [&text_streamer](std::vector<int64_t>&& tokens, LLMPipeline& pipe){
+        text_streamer.put(tokens[0]);
+    };
+
+    cout << "greedy generate streaming mode:" << endl;
+    config.max_new_tokens(20).set_callback(text_streamer_callback);
+    pipe(prompt, config);
+    text_streamer.end();
+    
+    // Example 2: Grouped Beam Search decoding example
+    pipe = LLMPipeline(model_path, device);  
+    config = pipe.generation_config();
+
+    // will return vector with num_return_sequences strings
+    auto num_return_sequences = 3;
+    config.max_new_tokens(20).num_groups(3).group_size(5).num_return_sequences(num_return_sequences);
+    
+    cout << endl << "grouped beam search generated candidates:" << endl;
+    auto generation_results = pipe({prompt}, config);
```

Comment:
Actually we need here `{}` because in order to get several sequences as output we need to call pipe with vector/initializer list input signature. 

So if we want to get several sequences by beam search we need to pass input prompt as a vector. But if we want just one sequence without scores then yes we can pass just prompt.